### PR TITLE
Speeds up table of contents load for casebook and section pages

### PR DESCRIPTION
**Pull request by varun-magesh**
Speeds up load by:
1. prefetching resources
2. selecting annotations in advance
3. using preloaded children tree

### CODE REVIEW

File: web/main/utils.py
Comment (rebeccacremona) [inline web/main/utils.py]: I think [this setup](https://github.com/harvard-lil/h2o/pull/1647/files#diff-520b40584c9ab08368ee82905be5a0fae5c23cf54c94c00516356f701ddf006dR759-R777) wants to become its own test fixture defined in [conftest.py](https://github.com/harvard-lil/h2o/blob/develop/web/conftest.py) and then summoned here by name. See [full_casebook_parts_factory](https://github.com/harvard-lil/h2o/blob/develop/web/conftest.py#L462) and [full_casebook](https://github.com/harvard-lil/h2o/blob/develop/web/conftest.py#L532).... or, does the existing `full_casebook` fixture do what you need? If not, I suggest when naming/making your fixture, including some comments that explain what kind of casebook structure you are looking for here.

File: web/main/utils.py
Comment (rebeccacremona) [inline web/main/utils.py]: I think some stray indentation.

File: web/main/utils.py
Comment (rebeccacremona) [inline web/main/utils.py]: Awesome. For extra clarity, that this is something special, maybe consider `has_annotation`->`_has_annotation`? Just a thought.

File: web/main/utils.py
Comment (rebeccacremona) [inline web/main/utils.py]: Does this need to be if `t.resource_type or t.resource_type == "Section" or t.resource_type == ""`, like above? I think yes: I think not all sections have `resource_type='Section'`, especially old ones. https://github.com/harvard-lil/h2o/issues/1448

File: web/main/utils.py
Comment (rebeccacremona) [inline web/main/utils.py]: The nitpickiest of nitpicks: here and below, I keep getting confused by the variable name `cns` here, thinking it means `content nodes`, or `content node serializer`, when in fact it is a single dict/already-serialized node. Maybe `node` or `node_dict` or `serialized_node`, or something else would be clearer?

File: web/main/utils.py
Comment (rebeccacremona) [inline web/main/utils.py]: I'll admit I'm struggling a little with how this bit works. Would using a fresh dict as the root, like the below, also work, or is it important to do the work on a falsey key in `serialized`?

```
tmp_root = {"children": []}
for ordinals, serialized_node in serialized.items():
    try:
        parent = serialized[ordinals[:-1]]
    except KeyError:
        # this node is at the top-level of the data being serialized
        parent = tmp_root
    try:
        parent["children"].append(serialized_node)
    except KeyError:
        raise ValueError("Trying to append children to non-Section!")
    return tmp_root["children"]
```
Is that the same?

File: web/main/serializers.py
Comment (rebeccacremona) [inline web/main/serializers.py:68]: True? This is identical to `SectionOutlineSerializer` except for `children = RecursiveField(many=True, allow_null=True, default=[])` -> `children = []` (and therefore no `children` in the `Meta`, and the presence of `ordinals` as well as `ordinal_string`?

If True, can `SectionOutlineSerializer` be removed entirely?

File: web/main/utils.py
Comment (rebeccacremona) [inline web/main/utils.py]: I think some stray indentation.

File: web/main/utils.py
Comment (rebeccacremona) [inline web/main/utils.py]: Same here: for extra clarity, maybe this special property should be `_transmutable`?

File: web/main/views.py
Comment (rebeccacremona) [inline web/main/views.py:574]: I don't quite understand this line; I think I'm stumbling over the `|`. Can you add a comment?

File: web/main/views.py
Comment (rebeccacremona) [inline web/main/views.py]: I have a syntax recommendation! To make it clear that you expect `manually_serialize_content_query(...)` to return a list of length one and extract that item at the same time, you can:
```
[mscq] = manually_serialize_content_query(....)
```
I think that would be a bit clearer than `[0]`, which could make a person wonder why we aren't returning the [non-existent) rest of the list too.

File: web/main/views.py
Comment (rebeccacremona) [inline web/main/views.py]: (same as above)

**Review by rebeccacremona — CHANGES_REQUESTED** [review]
This is SO FANTASTIC 🎉 Thanks so much for noticing and tackling this!

I have a bunch of little notes/questions made inline.

When we were chatting about this code and you mentioned not being sure where to put it, I mentioned `utils.py` as a catch-all: I think that actually, it would be extremely natural for this code to live in `serializers.py`... and I think that means you wouldn't have to do anything fancy with the imports.

I think you can just put the method in there, but now I'm musing whether it might be elegant to write it as class that looks more like a DRF serializer, something like:
```
class CustomFooSerializer:

  __init__(content_query: QuerySet):
    # logic of `manually_serialize_content_query` except...
    self.data = {...........}
```
It's probably too crazy to use https://www.django-rest-framework.org/api-guide/serializers/#baseserializer..... 

Maybe either class-based idea is goofy....

Again, thanks for this optimization!

**Comment by rebeccacremona** [comment]
Oh! I had one more thought: I think this deserves an `assert_num_queries` in the tests, to brag about how efficient it is.

File: web/main/utils.py
Comment (varun-magesh) [reply web/main/utils.py]: you mentioned this and I totally forgot --- will do!

File: web/main/utils.py
Comment (varun-magesh) [reply web/main/utils.py]: good to know!

File: web/main/utils.py
Comment (varun-magesh) [reply web/main/utils.py]: Yes, this is definitely the same --- the story here is that I originally wrote this to work for casebooks only, where all top-level sections have `len(ordinals)==1`. I didn't need a try catch there because the slice alone would take the ordinals to an empty tuple, and this solution felt rather natural. I came back later to modify it for sections, which required the try catch, and required me to give up the nice brevity of the original solution while keeping its extra confusing-ness.

**Comment by varun-magesh** [comment]
> It's probably too crazy to use https://www.django-rest-framework.org/api-guide/serializers/#baseserializer.....

I considered writing it as a DRF Serializer, but the problem I ran into was that the thing we are serializing  is not a Django model or object --- it's a QuerySet of a series of objects that we know how to put together. Even the previous implementation did some steps outside the serializer (converting a section's tree into a dict) and used the built-in serializer to serializer individual nodes in the tree. Basically (and I might be wrong because I haven't worked with serializers or django much), I think that what we are trying to do is different from what Serializers are made for (mainly classes and Django models).

So I think that using the serializer interface (either by copying it with `.data` or by inheriting explicitly) is misleading to someone who is familiar with serializers since this code does something pretty different than what serializers are expected to do --- one major example, it cannot be used to deserialize trees.

It seems like the best option might be to include this as a seperate method in serializers.py.

That being said, I'm not particularly familiar with Serializers or DRF; I wanted to explain my thought process here but if it makes more sense to use serializer-type objects here I'm happy to implement it that way.

**Comment by rebeccacremona** [comment]
> So I think that using the serializer interface (either by copying it with `.data` or by inheriting explicitly) is misleading to someone who is familiar with serializers since this code does something pretty different than what serializers are expected to do --- one major example, it cannot be used to deserialize trees.

👍 Fair enough

File: web/main/serializers.py
Comment (varun-magesh) [reply web/main/serializers.py:68]: True and True!

File: web/main/utils.py
Comment (varun-magesh) [reply web/main/utils.py]: full_casebook_factory is exactly what I needed, thankyou!

**Comment by varun-magesh** [comment]
Thanks so much for your feedback! Everything should be addressed now!

**Comment by varun-magesh** [comment]
Hmm --- before we merge this:

When deleting a section or a resource in casebook edit mode, it seems like this PR introduces a 1-2 second lag between hitting the "Delete" button and the section actually disappearing from view, which didn't use to happen. I'm not sure how this might be, I'd like to investigate a bit.

**Review by rebeccacremona — APPROVED** [review]
This looks great! Thanks so much for all the work!

We won't merge until told to, pending [your investigation about delete time/queries](https://github.com/harvard-lil/h2o/pull/1647#issuecomment-1194371492)

**Comment by varun-magesh** [comment]
Update: this bug is, unfortunately, very real. The requests actually look a little different before and after speedup. My code, however, is a drop-in replacement for the serialization that was happening before, and didn't touch requests, and in theory, the serialized data returned by views should be identical. Banged my head on this for too long, gonna take some air and go at this with a fresh pair of eyes tomorrow.

**Comment by varun-magesh** [comment]
Yay this is finally ready! The bug has been fixed. Thanks @rebeccacremona !

**Pull request merged**
